src/catchup/ApplyLedgerChainRange.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/CheckpointRange.h"
+
+namespace stellar
+{
+
+// Range of transaction checkpoints to apply.
+//
+// Transactions for checkpoins in checkpointRange() should be applied,
+// stoppint at ledger stopAt(). Ledger stopAt() must be in last checkpoint.
+class ApplyLedgerChainRange
```

Comment:
This code (in general, this class is part of the problem) is still quite hard to read as it still deals with too many types.

From what I see, we still have many range related classes:
 * CheckpointRange - (first, last, frequency)
 * ApplyLedgerChainRange - (mCheckpointRange, stopAt)
 * CatchupConfiguration - (lcl, to, count)
 * CatchupRange - (mApplyLedgerChain, applybuckets bool)

I don't think having all those objects makes sense as they are only a reflection of the state of how catchup is working, especially now that you have a centralized catchup implementation.

If you look at how catchup is implemented now.

CatchupWork, derives from BucketDownloadWork and spins up:
 * GetHistoryArchiveStateWork
 * BatchDownloadWork (checkpointrange )
 * VerifyLedgerChainWork (checkpointrange)
 * GetHistoryArchiveStateWork
 * DownloadBucketsWork
 * ApplyBucketsWork
 * BatchDownloadWork (checkpointrange)
 * ApplyLedgerChainWork(applyledgerchainrange)

From what I see, the only real benefit is to use CheckpointRange for work items that would otherwise fail. The only one that I can see that falls into this category is `CatchupWork` because of `BatchDownloadWork` 
 that has a strong dependency on checkpoint frequency.

The others (`VerifyLedgerChainWork` and `ApplyLedgerChainWork` ) should be able to work on an arbitrary range (the fact that it's lined up with checkpoints is only a concern for `CatchupWork`).

Now, as far as I can tell, the only reason you need all those types is that you want to ensure that the output of some calculation/work is properly used as the input to a different calculation/process at compile time but this is only a side effect of the way the code evolved: now that catchup is centralized you can have the code broken up in "steps" that can perform those computations on the fly using helper methods.

So here is what I am proposing:
* refactor the code to only use a minimum number of ledger number variables.
    * I think you only need to track the [`low`,`high`] being currently used by `CatchupWork` 
 (what is being downloaded/verified/applied etc) as well as the initial parameters (`toLedger` - that can be `CURRENT` and `count`)
    * a `CheckpointRange` should be computed in order to spin up `BatchDownloadWork` but is only used to update [low,high] at the `CatchupWork` level (it's basically a prep step before doing any work)
    * I do like that we have `CheckpointRange` as a parameter to BatchDownloadWork`
* all code should fall in place fairly easily as members of `CatchupWork`
* refactor the code such that `makeCatchupRange` becomes just an internal method to the `CatchupWork` class (or a few methods - some could be static so that you can continue to use the same tests that you wrote)